src/Quasar/Advanced/Types.purs

```diff
@@ -0,0 +1,393 @@
+      _, Just pt →
+        (map File $ lmap (const $ "Incorrect file resource") $ parseFile pt)
+        <|>
+        (map Dir $ lmap (const $ "Incorrect directory resource") $ parseDir pt)
+parseFile ∷ String → Either String (Pt.AbsFile Pt.Sandboxed)
+parseFile pt =
+  Pt.parseAbsFile pt
+  >>= Pt.sandbox Pt.rootDir
+  <#> (Pt.rootDir </> _)
+  # maybe (Left "Incorrect resource") pure
+
+parseDir ∷ String → Either String (Pt.AbsDir Pt.Sandboxed)
+parseDir pt =
+  Pt.parseAbsDir pt
+  >>= Pt.sandbox Pt.rootDir
+  <#> (Pt.rootDir </> _)
+  # maybe (Left "Incorrect resource") pure
+
+newtype Action =
+  Action
+    { operation ∷ Operation
+    , resource ∷ Resource
+    , resourceType ∷ ResourceType
+    }
+
+instance encodeJsonAction ∷ EncodeJson Action where
+  encodeJson (Action obj) =
+    "operation" := obj.operation
+    ~> "resource" := obj.resource
+    ~> "resourceType" := obj.resourceType
+    ~> jsonEmptyObject
+
+instance decodeJsonAction ∷ DecodeJson Action where
+  decodeJson = decodeJson >=> \obj →
+    { operation: _
+    , resource: _
+    , resourceType: _
+    }
+    <$> (obj .? "operation")
+    <*> (obj .? "resource")
+    <*> (obj .? "resourceType")
+    <#> Action
+
+
+newtype UserId = UserId String
+runUserId ∷ UserId → String
+runUserId (UserId s) = s
+
+instance encodeJsonUserId ∷ EncodeJson UserId where
+  encodeJson = runUserId >>> encodeJson
+
+instance decodeJsonUserId ∷ DecodeJson UserId where
+  decodeJson = map UserId <<< decodeJson
+
+
+newtype TokenId = TokenId String
+runTokenId ∷ TokenId → String
+runTokenId (TokenId s) = s
+
+instance encodeJsonTokenId ∷ EncodeJson TokenId where
+  encodeJson = runTokenId >>> encodeJson
+
+instance decodeJsonTokenId ∷ DecodeJson TokenId where
+  decodeJson = map TokenId <<< decodeJson
+
+newtype PermissionId = PermissionId String
+runPermissionId ∷ PermissionId → String
+runPermissionId (PermissionId s) = s
+
+instance encodeJsonPermissionId ∷ EncodeJson PermissionId where
+  encodeJson = runPermissionId >>> encodeJson
+
+instance decodeJsonPermissionId ∷ DecodeJson PermissionId where
+  decodeJson = map PermissionId <<< decodeJson
+
+
+data GrantedTo
+  = UserGranted UserId
+  | GroupGranted (Pt.AbsFile Pt.Sandboxed)
+  | TokenGranted TokenId
+
+instance encodeJsonGrantedTo ∷ EncodeJson GrantedTo where
+  encodeJson (UserGranted uid) = encodeJson uid
+  encodeJson (GroupGranted pt) = encodeJson $ Pt.printPath pt
+  encodeJson (TokenGranted tk) = encodeJson tk
+
+instance decodeJsonGrantedTo ∷ DecodeJson GrantedTo where
+  decodeJson = decodeJson >=> \str →
+    (map GroupGranted $ parseFile str)
+    <|>
+    (map (UserGranted <<< UserId) $ checkUserId str)
+    <|>
+    (pure $ TokenGranted $ TokenId str)
+    where
+    checkUserId ∷ String → Either String String
+    checkUserId str =
+      -- ¯\_(ツ)_/¯
+      if isJust (Str.indexOf "@" str)
+        then pure str
+        else Left "Incorrect email"
```

Comment:
Unfortunately not :( `grantedTo` field is `group|tokenId|userEmail`. It's better to add real email parser of course, but checking `@` should be enough for now I think.